.circleci/config.yml

```diff
@@ -351,12 +351,29 @@ defaults:
 
   - steps_install_dependencies_osx: &steps_install_dependencies_osx
       steps:
-        # FIXME: We used to cache dependencies on macOS but now it takes longer than just installing
-        # them each time. See https://github.com/ethereum/solidity/issues/12925.
+        - restore_cache:
+            keys:
+              - dependencies-osx-{{ arch }}-{{ checksum ".circleci/osx_install_dependencies.sh" }}
+        - attach_workspace:
+            at: .
+          # DO NOT EDIT between here and save_cache, but rather edit ./circleci/osx_install_dependencies.sh
+          # WARNING! If you do edit anything here instead, remember to invalidate the cache manually.
         - run:
             name: Install build dependencies
             command: ./.circleci/osx_install_dependencies.sh
 
+        - save_cache:
+            key: dependencies-osx-{{ arch }}-{{ checksum ".circleci/osx_install_dependencies.sh" }}
+            paths:
+              - /usr/local/bin
+              - /usr/local/lib
+              - /usr/local/include
+              - /usr/local/Cellar/boost
+              - /usr/local/Cellar/cmake
+              - /usr/local/Cellar/wget
+              - /usr/local/Cellar/coreutils
+              - /usr/local/Cellar/diffutils
```

Comment:
I wonder if it's really safe to do it like this. If you do not restore it all, we end up with some bits coming from cached homebrew and some coming from the image. What are the parts you are not caching?

This does seem to pass tests and gives a nice speedup so I'm a bit torn on this. Caching the whole `/usr/local/Cellar` and `/usr/local/Homebrew` like we used to seems like it would be less prone to break in weird ways at some point in the future.

Reply:
I honestly don't know if it is safe to do this way, but I decided to cache only what we need because if we cache everything inside the `Cellar` directory, it actually does not save us time. Restoring the cache ends up being slower (for example, see: https://app.circleci.com/pipelines/github/r0qs/solidity/640/workflows/b7416bf3-b054-48a1-ada1-d89c8e047a95/jobs/19678)

Reply:
What happens if we first remove everything from `/usr/local/Cellar` before restoring? My impression is that the slowdown really comes from printing all those errors (or maybe it's that just handling them goes through some horribly inefficient code path in `tar`?). It used to be pretty fast back when we used to do that.

Reply:
> What happens if we first remove everything from `/usr/local/Cellar` before restoring? 

Hum...not sure if this would be the best way, since it would "install" the packages in the `Cellar` twice in the first run (when there is no cache available). Also, it would cause errors like this: https://app.circleci.com/pipelines/github/ethereum/solidity/30938/workflows/1d8b9f77-a106-47a3-91d7-d561b4cfe8d2/jobs/1375906, i.e. some commands will not be available until we reinstall everything. I would rather just keep it as it is now, i.e. caching and restoring only the packages we need.

> My impression is that the slowdown really comes from printing all those errors (or maybe it's that just handling them goes through some horribly inefficient code path in `tar`?). It used to be pretty fast back when we used to do that.

Yes, I have the same impression.

Reply:
> Also, it would cause errors like this

But isn't that because you do not recreate an empty directory after deleting it?

> it would "install" the packages in the `Cellar` twice in the first run

And where's the problem with that? It's fine if the first run is longer as long as it's not extremely so. We update macOS dependencies quite rarely so the cache will almost always be present.

Reply:
> I would rather just keep it as it is now, i.e. caching and restoring only the packages we need.

I'd really prefer something we can just do and forget. Too many parts of our CI are built on weak assumptions and break all the time, distracting us. We could still go for it if the speedup was critical but at the moment it's more of an annoyance so maybe it would be better to live without caching until CircleCI fixes this on their side.

Reply:
> > Also, it would cause errors like this
> 
> But isn't that because you do not recreate an empty directory after deleting it?

No, I was referring to this error here:
```
/Users/distiller/.bash_profile: line 19: rbenv: command not found
/Users/distiller/.bash_profile: line 22: pyenv: command not found
Running `brew update --auto-update`...
```
I copied the wrong link in the previous comment: https://app.circleci.com/pipelines/github/ethereum/solidity/30938/workflows/1d8b9f77-a106-47a3-91d7-d561b4cfe8d2/jobs/1375906

> > it would "install" the packages in the `Cellar` twice in the first run
> 
> What's the problem with that? It's fine if the first run is longer as long as it's not extremely so. We update macOS dependencies quite rarely so the cache will almost always be present.

Well, I suspect it will be almost twice the time that it is currently, which is around 5 minutes.

Reply:
> I copied the wrong link in the previous comment: https://app.circleci.com/pipelines/github/ethereum/solidity/30938/workflows/1d8b9f77-a106-47a3-91d7-d561b4cfe8d2/jobs/1375906?invite=true#step-105-0_65

Looks like the same link to me, just an extra anchor.

> Well, I suspect it will be almost twice the time that it is currently, which is around 5 minutes.

That's still acceptable if it runs rarely.

Reply:
> > I copied the wrong link in the previous comment: https://app.circleci.com/pipelines/github/ethereum/solidity/30938/workflows/1d8b9f77-a106-47a3-91d7-d561b4cfe8d2/jobs/1375906?invite=true#step-105-0_65
> 
> Looks like the same link to me, just an extra anchor.

Yes, this is what I meant haha

> 
> > Well, I suspect it will be almost twice the time that it is currently, which is around 5 minutes.
> 
> That would be bad if we had to do it on every run, but still acceptable if it only runs rarely.

Ok, if you think it is acceptable, I will do the change then.